ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InputsAndEffects.scala

```diff
@@ -76,6 +78,57 @@ private[kvutils] object InputsAndEffects {
       parties.toList.sorted.map(partyStateKey)
     }
 
+    def addOrUpdateResolvedContractId(
```

Comment:
It updates the mapping that should really be the same for the same key within the same transaction, I think setting it only the first time is meant as an optimization but I don't think it's needed and I don't think it's actually faster. @kamil-da Am I understanding this right? In the name of simplicity I changed the name and simplified the implementation to simply set an entry in the map.

Reply:
I intentionally made this method to set the mapping only if there is no entry for a particular contract key. This combined with in-execution-order traversal makes that we record the initial state of a participant and how it resolved contract keys.
I'm not sure if the current implementation would work in case we have a transaction where a contract key is being modified and resolves initially to contract A, but at the end of the transaction resolves to contract B.

Reply:
I've just come to the same conclusion after seeing the tests failing and analyzing the cause, reverted the logic. Thanks for elaborating on this @kamil-da.

Reply:
Also, renamed the method in `addResolvedContractIdIfAbsent` which is clearer IMO.

Reply:
I could've put more comments on this which would save you some work. Sorry :)

Reply:
This is a great discussion which is going to be lost. Please write a test case which fails with the incorrect logic and passes with the correct logic, so that we don't accidentally un-revert it.

Reply:
That'd be great to have such a test case!

Reply:
Any exercise by key fails with the incorrect logic, including several `KVUtilsTransactionSpec` cases, but I'll add a minimal one.

Reply:
Added `be able to pre-execute an exercise by key`.